frontend/annotation_resolver.py

```diff
@@ -108,9 +112,23 @@ def resolve(self, annotation):
 
             # Get the args and return types
             args_annotations = self.resolve_args(args_and_return[0][1:-1])
-            args_types = [self.resolve(arg) for arg in args_annotations]
-            return_type = self.resolve(args_and_return[1])
+            args_types = [self.resolve(arg, solver) for arg in args_annotations]
+            return_type = self.resolve(args_and_return[1], solver)
 
             return self.z3_types.funcs[len(args_types)](*(args_types + [return_type]))
 
+        if annotation[:5] == "Union":
+            # Parse Union type
+            assert annotation[5] == "[" and annotation[-1] == "]"
+
+            # Get the types of the union args
+            args_annotations = self.resolve_args(annotation[6:-1])
+            args_types = [self.resolve(arg, solver) for arg in args_annotations]
+
+            result_type = solver.new_z3_const("union")
+            solver.add(Or([result_type == arg for arg in args_types]),
```

Comment:
This isn't really a union type though, is it? It says the type of the method is one of the options, and then Z3 will pick which of those it is. If we half-support things like that, we should be aware what we get from that, and there should definitely be a comment somewhere that explains what we support and what we don't.